Add tools script to list `torch` APIs

driazati opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58610 Add tools script to list `torch` APIs**

This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
yes | pip uninstall torch
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

# get 1.9.0 apis
yes | pip uninstall torch
pip install -f https://download.pytorch.org/whl/test/cu102/torch_test.html torch
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log
```



`diff` [`apis-1.8.1.log`](https://gist.github.com/2b97bbbea5fc4ec4b1451b6a8295ed3c) [`apis-1.9.0.log`](https://gist.github.com/be3f53cc3dc2596217021a0cd8e6130e) -> https://gist.github.com/12a83992268f63ebe8f3eaa37a5301f3

just the removed APIs: https://gist.github.com/ca06021f245754952af378b52a908ccf

driazati committed:
Add tools script to list `torch` APIs

This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

[ghstack-poisoned]

driazati committed:
Update on "Add tools script to list `torch` APIs"


This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

yes | pip uninstall torch

# get 1.9.0 apis
git checkout release/1.9
python setup.py develop
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

diff apis-1.8.1.log apis-1.9.0.log
```

[ghstack-poisoned]

driazati committed:
Update on "Add tools script to list `torch` APIs"


This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

yes | pip uninstall torch

# get 1.9.0 apis
git checkout release/1.9
python setup.py develop
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

diff apis-1.8.1.log apis-1.9.0.log
```

[ghstack-poisoned]

driazati committed:
Update on "Add tools script to list `torch` APIs"


This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
yes | pip uninstall torch
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

# get 1.9.0 apis
yes | pip uninstall torch
pip install -f https://download.pytorch.org/whl/test/cu102/torch_test.html torch
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log
```



`diff` [`apis-1.8.1.log`](https://gist.github.com/89e74ea51914eace0ed7fad64bc61116) [`apis-1.9.0.log`](https://gist.github.com/2cf3a68ef0bc13938937de6a8c4e6484) -> https://gist.github.com/64993c55f68e0ebd2b0a6e0509fb9c9a


[ghstack-poisoned]

albanD commented:
Do we want to organize the tools/ folder a bit better so that it doesn't become a flat dump of scripts?
Also if you just have a single script (and not a full tool like codegen), there is a script/ folder that might be better to put this.

walterddr commented on tools/list_apis.py:
```diff
@@ -0,0 +1,520 @@
+    print(*args, file=sys.stderr)
+
+
+system_module_names = {
```
Is there any better way other than listing out all system modules? for example using `modules` dunder methods

driazati replied:
I couldn't [find any](https://stackoverflow.com/questions/6463918/how-can-i-get-a-list-of-all-the-python-standard-library-modules), `isort` (the import ordering checker) does something similar: https://github.com/PyCQA/isort/tree/main/isort/stdlibs

What do you mean `modules` dunder methods?

walterddr replied:
I haven't dug deeper but out of top of my head we can check if `module.__file__` is coming from one of the site-packages other than of `site-package/torch`.

Or even simpler can we check if the module is coming from one of the torch file? Ultimately we are only interested in the torch module right?

albanD replied:
In case it helps, here is the version of this script that numpy uses: https://github.com/numpy/numpy/blob/main/numpy/tests/test_public_api.py

driazati committed:
Update on "Add tools script to list `torch` APIs"


This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
yes | pip uninstall torch
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

# get 1.9.0 apis
yes | pip uninstall torch
pip install -f https://download.pytorch.org/whl/test/cu102/torch_test.html torch
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log
```



`diff` [`apis-1.8.1.log`](https://gist.github.com/2b97bbbea5fc4ec4b1451b6a8295ed3c) [`apis-1.9.0.log`](https://gist.github.com/be3f53cc3dc2596217021a0cd8e6130e) -> https://gist.github.com/12a83992268f63ebe8f3eaa37a5301f3

just the removed APIs: https://gist.github.com/ca06021f245754952af378b52a908ccf

[ghstack-poisoned]

driazati committed:
Update on "Add tools script to list `torch` APIs"


This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
yes | pip uninstall torch
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

# get 1.9.0 apis
yes | pip uninstall torch
pip install -f https://download.pytorch.org/whl/test/cu102/torch_test.html torch
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log
```



`diff` [`apis-1.8.1.log`](https://gist.github.com/2b97bbbea5fc4ec4b1451b6a8295ed3c) [`apis-1.9.0.log`](https://gist.github.com/be3f53cc3dc2596217021a0cd8e6130e) -> https://gist.github.com/12a83992268f63ebe8f3eaa37a5301f3

just the removed APIs: https://gist.github.com/ca06021f245754952af378b52a908ccf

[ghstack-poisoned]

driazati committed:
Update on "Add tools script to list `torch` APIs"


This adds a script that will run through everything in `dir(torch)` recursively and print out each path. We can run this for 1.8.1 and then again for 1.9.0 to determine the API difference between the two in order to ensure everything is properly documented and there are no secret BC-breaking changes.

```bash
yes | pip uninstall torch
pip install torch
# get 1.8.1 apis
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log

# get 1.9.0 apis
yes | pip uninstall torch
pip install -f https://download.pytorch.org/whl/test/cu102/torch_test.html torch
python tools/list_apis.py --module torch > apis-$(python -c 'import torch; print(torch.__version__)').log
```



`diff` [`apis-1.8.1.log`](https://gist.github.com/2b97bbbea5fc4ec4b1451b6a8295ed3c) [`apis-1.9.0.log`](https://gist.github.com/be3f53cc3dc2596217021a0cd8e6130e) -> https://gist.github.com/12a83992268f63ebe8f3eaa37a5301f3

just the removed APIs: https://gist.github.com/ca06021f245754952af378b52a908ccf

[ghstack-poisoned]

albanD commented:
cc @anjali411

driazati commented:
abandoning in favor of #60927
